implementations/rust/ockam/ockam_api/src/authenticator/direct/direct_authenticator.rs

```diff
@@ -0,0 +1,155 @@
+use either::Either;
+use std::collections::{BTreeMap, HashMap};
+
+use ockam::identity::utils::now;
+use ockam::identity::AttributesEntry;
+use ockam::identity::Identifier;
+use ockam_core::compat::sync::Arc;
+use ockam_core::Result;
+
+use crate::authenticator::{
+    AuthorityMember, AuthorityMembersRepository, EnrollerAccessControlChecks,
+};
+
+/// Identity attribute key that indicates the role of the subject
+pub const OCKAM_ROLE_ATTRIBUTE_KEY: &str = "ockam-role";
+
+/// Identity attribute value that indicates the enroller role of the subject
+/// the corresponding key is [`OCKAM_ROLE_ATTRIBUTE_KEY`]
+pub const OCKAM_ROLE_ATTRIBUTE_ENROLLER_VALUE: &str = "enroller";
+
+pub struct DirectAuthenticatorErrorString(pub String);
+
+pub struct DirectAuthenticator {
+    members: Arc<dyn AuthorityMembersRepository>,
+}
+
+impl DirectAuthenticator {
+    pub fn new(members: Arc<dyn AuthorityMembersRepository>) -> Self {
+        Self { members }
+    }
+
+    pub async fn add_member(
+        &self,
+        enroller: &Identifier,
+        id: &Identifier,
+        attrs: &BTreeMap<String, String>,
+    ) -> Result<Either<(), DirectAuthenticatorErrorString>> {
+        let check =
+            EnrollerAccessControlChecks::check_identifier(self.members.clone(), enroller).await?;
+
+        if !check.is_enroller {
+            warn!("Not enroller {} is trying to add a member {}", enroller, id);
+            return Ok(Either::Right(DirectAuthenticatorErrorString(
+                "Not enroller is trying to add a member".to_string(),
+            )));
+        }
+
+        let attrs = attrs
+            .iter()
+            .map(|(k, v)| (k.as_bytes().to_vec(), v.as_bytes().to_vec()))
+            .collect();
+
+        // Check if we're trying to create an enroller
+        if EnrollerAccessControlChecks::check_bin_attributes_is_enroller(&attrs) {
+            // Only pre-trusted identities will be able to add enrollers
+            if !check.is_pre_trusted {
+                warn!(
+                    "Not pre trusted enroller {} is trying to create an enroller {}",
+                    enroller, id
+                );
+
+                return Ok(Either::Right(DirectAuthenticatorErrorString(
+                    "Not pre trusted enroller is trying to create an enroller".to_string(),
+                )));
+            }
+        }
+
+        let member = AuthorityMember::new(id.clone(), attrs, Some(enroller.clone()), now()?, false);
+
+        if let Err(err) = self.members.add_member(member).await {
+            warn!("Error adding member {} directly: {}", id, err);
+            return Ok(Either::Right(DirectAuthenticatorErrorString(
+                "Error adding member".to_string(),
```

Comment:
Why not return the `err` to the caller?

Reply:
I think we should avoid returning bare errors to the user in case of RPC. It's hard to guarantee that they are meaningful and not leak too much detail.